### PR TITLE
Add Part 03 Step 16

### LINKED ISSUE #285: Write Part - 3 Step - 16 & unit test
Implement Step16Controller and write all unit tests. Verify the display name of the test conforms to the pattern of "Part X Step Y". Follow the standardize naming, comments, etc. to be consistent.

### PR DESCRIPTION
**Pull request by mmschaefer**
Fix #285 - Add part 03 and step 16 to the project along with it associated unit tests.

### CODE REVIEW

File: src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java [hunk 1]
```diff
@@ -54,12 +58,41 @@
     @Override
     protected void run() throws Throwable {
         // 6.3.16.1.a. Turn the engine off.
+        incrementProgress("Part 3, Step 16 Turn Engine Off and keep the ignition key in the off position");
+        ensureKeyOffEngineOff();
+
         // 6.3.16.1.b. Confirm Fault A is still implanted according to the manufacturer’s instruction.
+        incrementProgress("Waiting for implant of Fault A according to the engine manufacturer's instruction");
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java]: This message shouldn't include "engine".  The spec doesn't indicate which module should have the fault.

Comment (gumbelmj) [reply src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java]: Actually, this message is wrong.  It should read:

"Please confirm Fault A is still implanted according to the manufacturer's instruction."

File: src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java [hunk 2]
```diff
@@ -54,12 +58,41 @@
...
+        waitForFault("Part 6.3.16.1.b");
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java]: This shouldn't wait for a fault.  It should just have the your press "OK or Cancel"

File: src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java [hunk 3]
```diff
@@ -54,12 +58,41 @@
...
+
         // 6.3.16.1.c. Wait manufacturer’s recommended interval with the key in the off position.
+        incrementProgress("Waiting for manufacturer's recommended interval after the Fault A implant");
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java:69]: This appears to be copy pasted from the previous step.

File: src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java [hunk 4]
```diff
@@ -54,12 +58,41 @@
...
+        }
+    }
+
+    private void waitForEngineStart() {
+        String message = "Turn ignition key to the ON position" + NL;
+        message += "Please observe the MIL and Wait to Start Lamp (if equipped) in the Instrument Cluster" + NL;
+        message += "Start Engine after MIL and Wait to Start Lamp (if equipped) have extinguished" + NL;
+        message += "Please wait as indicated by the engine manufacturer’s recommendations for Fault A" + NL;
+        message += "Press OK when ready to continue testing" + NL;
+        if (!isDevEnv()) {
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java:97]: nit: this `isDevEnv()` check can happen before the `String` is created as the string will never be shown so it doesn't need to be created.  You might even be able to move the check up higher.

File: src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java (hunk 1 above)
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java]: I noticed the first `incrementProgress()` has the Part and Step number before the message.  I like that.  I think we should do that in all cases where possible.

File: src/org/etools/j1939_84/controllers/StepController.java [hunk 5]
```diff
@@ -270,7 +270,9 @@ protected void checkForNACKsFromObdModules(List<? extends GenericPacket> packets
     protected void waitForFault(String boxTitle) {
         String message = "Implant Fault A according to engine manufacturer’s instruction" + NL;
         message += "Press OK when ready to continue testing" + NL;
-        displayInstructionAndWait(message, boxTitle, WARNING);
+        if(!isDevEnv()) {
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/StepController.java:273]: See comment about moving the `isDevEnv()` method out/up.

**Review by gumbelmj — CHANGES_REQUESTED** [review]
A few minor points.

File: src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java (hunk 3 above)
Comment (mmschaefer) [reply src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java:69]: What would you like this to say?

File: src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java (hunk 2 above)
Comment (mmschaefer) [reply src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java]: Per your reqeust:
 // 6.3.16.1.b. Confirm Fault A is still implanted according to the manufacturer’s instruction.
        incrementProgress("Please confirm Fault A is still implanted according to the manufacturer's instruction");
        waitForFault("Part 6.3.16.1.b");

File: src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java (hunk 3 above)
Comment (gumbelmj) [reply src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java:69]: Perhaps "Part 3, Step 16: Waiting manufacturer’s recommended interval with the key in the off position."  This will be the message displayed in the status bar while the dialog box is shown.  If they take a screen shot, it should reflect what the app is actually doing.

"Part 3, Step 16" could be "6.3.16.1.c", but I think we should be consistent.. at least going forward from here.

File: src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java (hunk 2 above)
Comment (gumbelmj) [reply src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java]: This Step (unlike the last) isn't waiting for a fault to be implanted.  It's waiting per the manufacturer's instructions.

File: src/org/etools/j1939_84/controllers/part03/Part03Step14Controller.java [hunk 6]
```diff
@@ -54,7 +60,19 @@
     @Override
     protected void run() throws Throwable {
         // 6.3.14.1.a. DS DM20 (send Request (PGN 59904) for PGN 49664 (SPNs 3048)) to ECU(s) that responded in part 1 with DM20 data.
+        List<BusResult<DM20MonitorPerformanceRatioPacket>> dsResults = getDataRepository().getObdModuleAddresses()
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part03/Part03Step14Controller.java]: Why is this here?

File: src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java (hunk 2 above)
Comment (gumbelmj) [reply src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java]: This needs to display an OK/Cancel dialog indicating we are waiting according to the manufacturer's instructions.

**Review by gumbelmj — CHANGES_REQUESTED** [review]
The wrong dialog is being shown.

File: src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java (hunk 2 above)
Comment (mmschaefer) [reply src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java]: That is c.
 // 6.3.16.1.c. Wait manufacturer’s recommended interval with the key in the off position.
        incrementProgress("Waiting for manufacturer's recommended interval after the Fault A implant");
        waitForManufacturerInterval();

Comment (gumbelmj) [reply src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java]: 6.3.16.1.b is about the user confirming the fault is still implanted, not waiting.

The `waitForFault()` method will display:
> Implant Fault A according to engine manufacturer’s instruction
> Press OK when ready to continue testing

That's not what we asking the user to do at this point.  We are asking them to confirm the fault still exists.

File: src/org/etools/j1939_84/controllers/part03/Part03Step14Controller.java (hunk 6 above)
Comment (gumbelmj) [reply src/org/etools/j1939_84/controllers/part03/Part03Step14Controller.java]: This code shouldn't be in here.

**Review by gumbelmj — CHANGES_REQUESTED** [review]
The message displayed to the user needs updated and the Part 3 Step 14 code shouldn't be in here.

File: src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java [hunk 7]
```diff
@@ -54,12 +58,50 @@
...
+
         // 6.3.16.1.d. Turn ignition key to the ON position.
         // 6.3.16.1.e. Observe MIL and Wait to Start Lamp in Instrument Cluster
         // 6.3.16.1.f. Start Engine after MIL and Wait to Start Lamp (if equipped) have extinguished.
         // 6.3.16.1.g. Wait as indicated by the engine manufacturer’s recommendations for Fault A.
+        waitForEngineStart();
+    }
+
+    private void confirmFault(){
+        String message = "Please confirm Fault A has been implanted" + NL;
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java:80]: nit: "Please confirm Fault A is still implanted"

File: src/org/etools/j1939_84/controllers/part03/Part03Step14Controller.java (hunk 6 above)
Comment (gumbelmj) [reply src/org/etools/j1939_84/controllers/part03/Part03Step14Controller.java]: Step 14 still needs to be removed.

**Review by gumbelmj — CHANGES_REQUESTED** [review]
Step14 needs to be removed.

File: src/org/etools/j1939_84/controllers/part03/Part03Step14Controller.java [hunk 8]
```diff
@@ -7,6 +7,7 @@
 import java.util.concurrent.Executors;
 import org.etools.j1939_84.controllers.DataRepository;
 import org.etools.j1939_84.controllers.StepController;
+import org.etools.j1939_84.model.OBDModuleInformation;
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part03/Part03Step14Controller.java]: This line needs removed.

**Review by gumbelmj — CHANGES_REQUESTED** [review]
Remove import

**Review by gumbelmj — APPROVED** [review]
